File: ProjectCodes/RProject/src/eda.R
library(R6)
library(dplyr)
library(ggplot2)

EDAAnalyzer <- R6Class("EDAAnalyzer",
  public = list(
    df = NULL,
    config = NULL,
    numeric_cols_summary = NULL,
    numeric_cols_hist = NULL,
    scatter_plots = NULL,
    output_dir = NULL,

    # -------------------- INIT --------------------
    initialize = function(df, config) {
      self$df <- df
      self$config <- config
      self$numeric_cols_summary <- config$eda$numeric_summary
      self$numeric_cols_hist <- config$eda$numeric_hist
      self$scatter_plots <- config$eda$scatter_plots
      self$output_dir <- config$output_dir_r
      dir.create(self$output_dir, showWarnings = FALSE, recursive = TRUE)
    },

    # -------------------- BASIC SUMMARY STATS --------------------
    summary_statistics = function() {
      cat("\nBasic statistics:\n")
      print(summary(self$df[, self$numeric_cols_summary]))
    },

    # -------------------- TOP AIRLINES --------------------
    plot_top_airlines = function(top_n = 10) {
      airline_counts <- self$df %>%
        count(AIRLINE, sort = TRUE) %>%
        slice_head(n = top_n)

      p <- ggplot(airline_counts, aes(x = n, y = reorder(AIRLINE, n))) +
        geom_bar(stat = "identity", fill = "steelblue") +
        labs(
          title = paste("Top", top_n, "Airlines by Number of Flights"),
          x = "Count",
          y = "Airline"
        ) +
        theme_minimal()

      ggsave(file.path(self$output_dir, "top_airlines_count.png"), plot = p,
             width = 12, height = 6, bg = "white")
      cat("Top airlines plot saved.\n")
    },

    # -------------------- HISTOGRAM PLOTS --------------------
    plot_histograms = function() {
      for (col in self$numeric_cols_hist) {
        if (!col %in% colnames(self$df)) next

        p <- ggplot(self$df, aes(x = .data[[col]])) +
          geom_histogram(aes(y = after_stat(density)), bins = 50,
                         fill = "steelblue", color = "white", alpha = 0.7) +
          geom_density(color = "red", linewidth = 0.8) +
          labs(title = paste("Distribution of", col), x = col, y = "Density") +
          theme_minimal()

        ggsave(file.path(self$output_dir, paste0("hist_", col, ".png")), plot = p,
               width = 8, height = 4, bg = "white")
      }
      cat("Histograms saved.\n")
    },

    # -------------------- BOXPLOT --------------------
    plot_boxplots_by_airline = function(target_cols = c("ARR_DELAY")) {
      for (col in target_cols) {
        if (!col %in% colnames(self$df)) next

        p <- ggplot(self$df, aes(x = AIRLINE, y = .data[[col]])) +
          geom_boxplot(fill = "steelblue", alpha = 0.7, outlier.alpha = 0.1) +
          coord_cartesian(ylim = c(-100, 100)) +
          labs(title = paste("Boxplot of", col, "by Airline"),
               x = "Airline", y = col) +
          theme_minimal() +
          theme(axis.text.x = element_text(angle = 45, hjust = 1))

        ggsave(file.path(self$output_dir, paste0("boxplot_", col, "_by_airline.png")),
               plot = p, width = 12, height = 6, bg = "white")
      }
      cat("Boxplots saved.\n")
    },

    # -------------------- FEATURES CORRELATION --------------------
    plot_correlation_heatmap = function() {
      # Select only existing numeric cols
      cols <- self$numeric_cols_hist[self$numeric_cols_hist %in% colnames(self$df)]
      corr_matrix <- cor(self$df[, cols], use = "complete.obs")

      # Melt for ggplot
      corr_df <- as.data.frame(as.table(corr_matrix))
      colnames(corr_df) <- c("Var1", "Var2", "Correlation")

      p <- ggplot(corr_df, aes(x = Var1, y = Var2, fill = Correlation)) +
        geom_tile() +
        geom_text(aes(label = round(Correlation, 2)), size = 2.5) +
        scale_fill_gradient2(low = "blue", mid = "white", high = "red", midpoint = 0) +
        labs(title = "Correlation Heatmap") +
        theme_minimal() +
        theme(axis.text.x = element_text(angle = 45, hjust = 1))

      ggsave(file.path(self$output_dir, "corr_heatmap.png"), plot = p,
             width = 12, height = 10, bg = "white")
      cat("Correlation heatmap saved.\n")
    },

    # -------------------- KDE PLOT --------------------
    plot_kde_by_airline = function(target_cols = c("ARR_DELAY")) {
      for (col in target_cols) {
        if (!col %in% colnames(self$df)) next

        p <- ggplot(self$df, aes(x = .data[[col]], fill = AIRLINE, color = AIRLINE)) +
          geom_density(alpha = 0.4) +
          labs(title = paste("KDE Plot of", col, "by Airline"),
               x = col, y = "Density") +
          theme_minimal()

        ggsave(file.path(self$output_dir, paste0("kde_", col, "_by_airline.png")),
               plot = p, width = 12, height = 6, bg = "white")
      }
      cat("KDE plots saved.\n")
    },

    # -------------------- SCATTER PLOTS --------------------
    plot_scatter = function(scatter_configs = NULL) {
      if (is.null(scatter_configs)) {
        scatter_configs <- self$scatter_plots
      }

      for (entry in scatter_configs) {
        x_col    <- entry[[1]]
        y_col    <- entry[[2]]
        title    <- entry[[3]]
        filename <- entry[[4]]

        if (!x_col %in% colnames(self$df) || !y_col %in% colnames(self$df)) next

        # Sample for performance
        plot_df <- self$df %>% sample_n(min(50000, nrow(self$df)))

        p <- ggplot(plot_df, aes(x = .data[[x_col]], y = .data[[y_col]])) +
          geom_point(alpha = 0.3, size = 0.5, color = "steelblue") +
          labs(title = paste("Scatter Plot:", title), x = x_col, y = y_col) +
          theme_minimal()

        ggsave(file.path(self$output_dir, filename), plot = p,
               width = 8, height = 6, bg = "white")
      }
      cat("Scatter plots saved.\n")
    },

    # -------------------- RUN ALL --------------------
    run_all = function() {
      cat("\n", strrep("=", 20), " EDA STEP ", strrep("=", 20), "\n")

      cat("\n1. Computing summary statistics...\n")
      self$summary_statistics()

      cat("2. Plotting top airlines by flight count...\n")
      self$plot_top_airlines()

      cat("3. Plotting histograms for numeric columns...\n")
      self$plot_histograms()

      cat("4. Plotting boxplots of ARR_DELAY by airline...\n")
      self$plot_boxplots_by_airline()

      cat("5. Plotting correlation heatmap for numeric features...\n")
      self$plot_correlation_heatmap()

      cat("6. Plotting KDE of ARR_DELAY by airline...\n")
      self$plot_kde_by_airline()

      cat("7. Generating scatter plots as per config...\n")
      self$plot_scatter()

      cat("EDA complete. All outputs saved to:", self$output_dir, "\n")
    }
  )
)

File: ProjectCodes/PythonProject/src/eda.py
import pandas as pd
import seaborn as sns
import matplotlib.pyplot as plt
import warnings
import os

warnings.simplefilter(action='ignore', category=FutureWarning)

class EDAAnalyzer:
    def __init__(self, df: pd.DataFrame, config: dict):
        self.df = df.copy()
        self.config = config
        self.numeric_cols_summary = config['eda']['numeric_summary']
        self.numeric_cols_hist = config['eda']['numeric_hist']
        self.scatter_plots = config['eda']['scatter_plots']
        self.output_dir = config['output_dir']
        os.makedirs(self.output_dir, exist_ok=True)

        self.post_fe_hist_cols = config['eda_post_engineering']['numeric_hist']
        self.post_fe_corr_cols = config['eda_post_engineering']['correlation_cols']
        self.post_fe_boxplot_cols = config['eda_post_engineering']['boxplot']
        self.output_dir_post_fe = config['output_dir_post_fe']
        os.makedirs(self.output_dir_post_fe, exist_ok=True)

    # -------------------- BASIC SUMMARY STATS --------------------
    def summary_statistics(self):
        """Print basic stats, median, mean, std for numeric columns."""
        print("\nBasic statistics:\n", self.df[self.numeric_cols_summary].describe())

    # -------------------- TOP AIRLINES --------------------
    def plot_top_airlines(self, top_n=10):
        """Count plot for top N airlines."""
        plt.figure(figsize=(12,6))
        sns.countplot(
            y='AIRLINE',
            data=self.df,
            order=self.df['AIRLINE'].value_counts().index[:top_n]
        )
        plt.title(f"Top {top_n} Airlines by Number of Flights")
        plt.xlabel("Count")
        plt.ylabel("Airline")
        plt.savefig(os.path.join(self.output_dir, "top_airlines_count.png"), bbox_inches='tight')
        plt.close()

    # -------------------- HISTOGRAM PLOTS --------------------
    def plot_histograms(self):
        """Plot histograms for numeric columns."""
        for col in self.numeric_cols_hist:
            plt.figure(figsize=(8,4))
            sns.histplot(self.df[col], bins=50, kde=True)
            plt.title(f"Distribution of {col}")
            plt.savefig(os.path.join(self.output_dir, f"hist_{col}.png"), bbox_inches='tight')
            plt.close()

    # -------------------- BOXPLOT --------------------
    def plot_boxplots_by_airline(self, target_cols=['ARR_DELAY']):
        """Boxplots of target columns by airline."""
        for col in target_cols:
            plt.figure(figsize=(12,6))
            sns.boxplot(x='AIRLINE', y=col, data=self.df)
            plt.ylim(-100, 100)
            plt.xticks(rotation=45)
            plt.savefig(os.path.join(self.output_dir, f"boxplot_{col}_by_airline.png"), bbox_inches='tight')
            plt.close()

    # -------------------- FEATURES CORRELATION --------------------
    def plot_correlation_heatmap(self):
        """Correlation matrix & heatmap for numeric columns."""
        corr_matrix = self.df[self.numeric_cols_hist].corr()
        plt.figure(figsize=(12,4))
        sns.heatmap(corr_matrix, annot=True, cmap='coolwarm')
        plt.title("Correlation Heatmap")
        plt.savefig(os.path.join(self.output_dir, "corr_heatmap.png"), bbox_inches='tight')
        plt.close()

    # -------------------- KDE PLOT --------------------
    def plot_kde_by_airline(self, target_cols=['ARR_DELAY']):
        """KDE plots of target columns per airline."""
        for col in target_cols:
            plt.figure(figsize=(12,6))
            sns.kdeplot(data=self.df, x=col, hue='AIRLINE', fill=True, alpha=0.5)
            plt.title(f"KDE Plot of {col} by Airline")
            plt.xlabel(col)
            plt.ylabel('Density')
            plt.savefig(os.path.join(self.output_dir, f"kde_{col}_by_airline.png"), bbox_inches='tight')
            plt.close()

    # -------------------- SCATTER PLOTS --------------------
    def plot_scatter(self, scatter_configs=None):
        """Generate scatter plots as per config list."""
        if scatter_configs is None:
            scatter_configs = self.scatter_plots

        for x_col, y_col, title, filename in scatter_configs:
            plt.figure(figsize=(8,6))
            sns.scatterplot(x=x_col, y=y_col, data=self.df, alpha=0.3)
            plt.title(f"Scatter Plot: {title}")
            plt.xlabel(x_col)
            plt.ylabel(y_col)
            plt.savefig(os.path.join(self.output_dir, filename), bbox_inches='tight')
            plt.close()

    # -------------------- POST FEATURE ENGINEERING EDA --------------------
    def plot_post_fe_histograms(self):
        """Plot histograms for engineered feature columns."""
        for col in self.post_fe_hist_cols:
            if col not in self.df.columns:
                print(f"⚠️  Column '{col}' not found in dataset, skipping...")
                continue
            plt.figure(figsize=(8, 4))
            sns.histplot(self.df[col], bins=30, kde=True)
            plt.title(f"Distribution of {col} (Engineered)")
            plt.xlabel(col)
            plt.ylabel("Count")
            plt.savefig(os.path.join(self.output_dir_post_fe, f"hist_eng_{col}.png"), bbox_inches='tight')
            plt.close()

    def plot_post_fe_correlation_heatmap(self):
        """Correlation heatmap for engineered features vs ARR_DELAY."""
        valid_cols = [c for c in self.post_fe_corr_cols if c in self.df.columns]
        if not valid_cols:
            print("⚠️  No valid columns found for post-FE correlation heatmap, skipping...")
            return
        corr_matrix = self.df[valid_cols].corr()
        plt.figure(figsize=(12, 6))
        sns.heatmap(corr_matrix, annot=True, cmap='coolwarm')
        plt.title("Correlation Heatmap - Engineered Features")
        plt.savefig(os.path.join(self.output_dir_post_fe, "corr_heatmap_engineered.png"), bbox_inches='tight')
        plt.close()

    def plot_post_fe_boxplots(self):
        """Boxplots of ARR_DELAY grouped by key engineered binary/categorical features."""
        for col in self.post_fe_boxplot_cols:
            if col not in self.df.columns or 'ARR_DELAY' not in self.df.columns:
                print(f"⚠️  Column '{col}' or 'ARR_DELAY' not found, skipping boxplot...")
                continue
            plt.figure(figsize=(8, 5))
            sns.boxplot(x=col, y='ARR_DELAY', data=self.df)
            plt.title(f"ARR_DELAY by {col}")
            plt.xlabel(col)
            plt.ylabel("Arrival Delay (min)")
            plt.savefig(os.path.join(self.output_dir_post_fe, f"boxplot_arrdelay_by_{col}.png"), bbox_inches='tight')
            plt.close()

    def run_post_engineering(self):
        """Run EDA on engineered features after feature engineering step."""
        print("\n" + "=" * 20 + " POST-ENGINEERING EDA " + "=" * 20)

        print("\n1️⃣  Plotting histograms for engineered features...")
        self.plot_post_fe_histograms()

        print("2️⃣  Plotting correlation heatmap for engineered features...")
        self.plot_post_fe_correlation_heatmap()

        print("3️⃣  Plotting ARR_DELAY boxplots by engineered categorical features...")
        self.plot_post_fe_boxplots()

        print("🎉 Post-engineering EDA complete. All outputs saved to:", self.output_dir_post_fe)

    # -------------------- RUN ALL --------------------
    def run_all(self):
        """Run all EDA plots and stats with clear prints for each step."""
        print("\n" + "=" * 20 + " EDA STEP " + "=" * 20)

        print("\n1️⃣  Computing summary statistics...")
        self.summary_statistics()

        print("2️⃣  Plotting top airlines by flight count...")
        self.plot_top_airlines()

        print("3️⃣  Plotting histograms for numeric columns...")
        self.plot_histograms()

        print("4️⃣  Plotting boxplots of ARR_DELAY by airline...")
        self.plot_boxplots_by_airline()

        print("5️⃣  Plotting correlation heatmap for numeric features...")
        self.plot_correlation_heatmap()

        print("6️⃣  Plotting KDE of ARR_DELAY by airline...")
        self.plot_kde_by_airline()

        print("7️⃣  Generating scatter plots as per config...")
        self.plot_scatter()

        print("🎉 EDA complete. All outputs saved to:", self.output_dir)